cpp/include/cudf/join.hpp

```diff
@@ -858,6 +859,310 @@ std::unique_ptr<rmm::device_uvector<size_type>> conditional_left_anti_join(
+ * using the corresponding mixed_full_join_size API).
+ * @param mr Device memory resource used to allocate the returned table and columns' device memory
+ *
+ * @return A pair of vectors [`left_indices`, `right_indices`] that can be used to construct
+ * the result of performing a mixed full join between the four input tables.
+ */
+std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+mixed_full_join(table_view const& left_equality,
+                table_view const& right_equality,
+                table_view const& left_conditional,
+                table_view const& right_conditional,
+                ast::expression const& binary_predicate,
+                null_equality compare_nulls = null_equality::EQUAL,
+                std::optional<std::pair<std::size_t, column_view>> output_size_data = {},
+                rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+
+/**
+ * @brief Returns the exact number of matches (rows) when performing a
+ * mixed inner join between the specified tables where the columns of the
+ * equality table are equal and the predicate evaluates to true on the
+ * conditional tables.
+ *
+ * If the provided predicate returns NULL for a pair of rows (left, right),
+ * that pair is not included in the output. It is the user's responsiblity to
+ * choose a suitable compare_nulls value AND use appropriate null-safe
+ * operators in the expression.
+ *
+ * @throw cudf::logic_error If the binary predicate outputs a non-boolean result.
+ * @throw cudf::logic_error If the number of rows in left_equality and left_conditional do not
+ * match.
+ * @throw cudf::logic_error If the number of rows in right_equality and right_conditional do not
+ * match.
+ *
+ * @param left_equality The left table used for the equality join.
+ * @param right_equality The right table used for the equality join.
+ * @param left_conditional The left table used for the conditional join.
+ * @param right_conditional The right table used for the conditional join.
+ * @param binary_predicate The condition on which to join.
+ * @param compare_nulls Whether or not null values join to each other or not.
+ * @param output_size An optional pair of values indicating the exact output size and the number of
+ * matches for each row in the larger of the two input tables, left or right (may be precomputed
+ * using the corresponding mixed_inner_join_size API).
+ * @param mr Device memory resource used to allocate the returned table and columns' device memory
+ *
+ * @return A pair containing the size that would result from performing the
+ * requested join and the number of matches for each row in one of the two
+ * tables. Which of the two tables is an implementation detail and should not
+ * be relied upon, simply passed to the corresponding `mixed_inner_join` API as
+ * is.
+ */
+std::pair<std::size_t, std::unique_ptr<column>> mixed_inner_join_size(
```

Comment:
Can we do this instead? 
```suggestion
std::pair<std::size_t, device_uvector<size_type>> mixed_inner_join_size(
```

Reply:
Using a `column` instead of a `device_uvector` was specifically requested by @jlowe. I told him that it's possible this API might change a little bit over the next release or two as we hammer out the details, but assuming that we do end up having the size kernels return the matches per row as we are doing now, the Spark plugin could make use of these sizes to more intelligently distribute the work across different workers. In that case making this return a column would make it easier for them to work with using existing libcudf algorithms that operate on columns.

Reply:
> I don't quite buy the argument that returning a unique_ptr<column> makes it all that much easier.

We don't have Java bindings for `device_uvector`, so in that sense it is a _ton_ easier for me. :smile:  If you want to keep the C++ API returning a `device_uvector` with the promise that it will always contain 2^31 elements or less, then I'll still expose this in Java as returning a `ColumnVector` (and fix it up in the JNI to build a `cudf::column` from the buffer) since Java can't really do much useful with `device_uvector` today.  Either that or I won't expose it to Java at all, at least for now, since currently it's opaque what it means and can be used for.

Reply:
In discussions offline we decided to stick with the `device_uvector` and have the JNI translate it into a column at the point of call.